<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="leg_robot">
    <link name="world"/>
    
    <joint name="base_joint" type="fixed">
        <parent link="world"/>
        <child link="base_link"/>
        <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
    </joint>
    <link name="base_link">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <mass value="1.0"/>
            <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
        </inertial>
        <visual>
            <origin xyz="0.0 -0.015 0.45" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="0.09 0.03 0.9"/>
            </geometry>
            <material name="red">
                <color rgba="1.0 0.0 0.0 2.0"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0.0 -0.015 0.45" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="0.09 0.03 0.9"/>
            </geometry>
        </collision>
    </link>
    <gazebo reference="base_link">
        <material>Gazebo/Red</material>
    </gazebo>

    <joint name="body_joint" type="fixed">
        <parent link="base_link"/>
        <child link="body_link"/>
        <origin xyz="0.0 0.0 0.7" rpy="0.0 0.0 0.0"/>
    </joint>
    <link name="body_link">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <mass value="1.0"/>
            <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 -0.7" rpy="0.0 0.0 ${radians(180)}"/>
            <geometry>
                <mesh filename="package://lily_dog/stl/leg_tester/body_link.stl"/>
            </geometry>
            <material name="green">
                <color rgba="0.0 1.0 0.0 2.0"/>
            </material>
        </visual>
        <collision>
            <origin xyz="-0.0175 0.0185 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="0.175 0.037 0.2"/>
            </geometry>
        </collision>
    </link>
    <gazebo reference="body_link">
        <material>Gazebo/Green</material>
    </gazebo>
    
    <joint name="shoulder1_joint" type="revolute">
        <parent link="body_link"/>
        <child link="shoulder_link"/>
        <axis xyz="0 0 1"/>
        <limit lower="${radians(-90)}" upper="${radians(90)}" velocity="50.0" effort="1000.0"/>
        <origin xyz="0 0.092 0" rpy="0 ${radians(90)} 0"/>
        <dynamics damping="1.0"/>
    </joint>
    <transmission name="shoulder1_trans" type="SimpleTransmission">
        <type>transmission_interface/SimpleTransmission</type>
        <actuator name="shoulder1_motor">
            <mechanismReduction>1</mechanismReduction>
        </actuator>
        <joint name="shoulder1_joint">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
    </transmission>
    <link name="shoulder_link">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <mass value="0.1"/>
            <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 ${radians(180)}"/>
            <geometry>
                <mesh filename="package://lily_dog/stl/leg_tester/shoulder_link.stl"/>
            </geometry>
            <material name="blue">
                <color rgba="0.0 0.0 1.0 2.0"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0.0 0.0 ${0.07-0.203/2}" rpy="0.0 0.0 0.0"/>
                <geometry>
                    <cylinder radius="0.04" length="0.203"/>
                </geometry>
        </collision>
    </link>
    <gazebo reference="shoulder_link">
        <material>Gazebo/Blue</material>
    </gazebo>
    
    <joint name="shoulder2_joint" type="continuous">
        <parent link="shoulder_link"/>
        <child link="upper_leg_link"/>
        <axis xyz="0 0 1"/>
        <limit velocity="50.0" effort="1000.0"/>
        <origin xyz="0 0.079 0" rpy="${radians(-90)} 0 0"/>
        <dynamics damping="1.0"/>
    </joint>
    <transmission name="shoulder2_trans" type="SimpleTransmission">
        <type>transmission_interface/SimpleTransmission</type>
        <actuator name="shouder2_motor">
            <mechanismReduction>1</mechanismReduction>
        </actuator>
        <joint name="shoulder2_joint">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
    </transmission>
    <link name="upper_leg_link">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <mass value="0.1"/>
            <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <mesh filename="package://lily_dog/stl/leg_tester/upper_leg_link.stl"/>
            </geometry>
            <material name="magenta">
                <color rgba="1.0 0.0 1.0 2.0"/>
            </material>
        </visual>
        <collision>
            <origin xyz="${0.388/2-0.05} 0.0 0.0" rpy="0.0 0.0 0.0"/>
                <geometry>
                    <box size="0.388 0.1 0.069"/>
                </geometry>
        </collision>
    </link>
    <gazebo reference="upper_leg_link">
        <material>Gazebo/Purple</material>
    </gazebo>

    <joint name="knee_joint" type="revolute">
        <parent link="upper_leg_link"/>
        <child link="lower_leg_link"/>
        <axis xyz="0 0 1"/>
        <limit lower="${radians(-150)}" upper="${radians(150)}" velocity="50.0" effort="1000.0"/>
        <origin xyz="0.3 0 0" rpy="${radians(180)} 0 0"/>
        <dynamics damping="1.0"/>
    </joint>
    <transmission name="knee_trans" type="SimpleTransmission">
        <type>transmission_interface/SimpleTransmission</type>
        <actuator name="knee_motor">
            <mechanismReduction>${60/44}</mechanismReduction>
        </actuator>
        <joint name="knee_joint">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
    </transmission>
    <link name="lower_leg_link">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <mass value="0.1"/>
            <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <mesh filename="package://lily_dog/stl/leg_tester/lower_leg_link.stl"/>
            </geometry>
            <material name="cyan">
                <color rgba="0.0 1.0 1.0 2.0"/>
            </material>
        </visual>
        <collision>
            <origin xyz="${0.34/2-0.0375} 0.0 0.0" rpy="0.0 0.0 0.0"/>
                <geometry>
                    <box size="0.34 0.06 0.026"/>
                </geometry>
        </collision>
    </link>
    <gazebo reference="lower_leg_link">
        <material>Gazebo/SkyBlue</material>
    </gazebo>

    <joint name="toe_joint" type="fixed">
        <parent link="lower_leg_link"/>
        <child link="toe_link"/>
        <origin xyz="0.3 0.0 0.0" rpy="0.0 0.0 0.0"/>
    </joint>
    <link name="toe_link">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <mass value="0.01"/>
            <inertia ixx="0.1" ixy="0.0" ixz="0.0" iyy="0.1" iyz="0.0" izz="0.1"/>
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <sphere radius="0.025"/>
            </geometry>
            <material name="yellow">
                <color rgba="1.0 1.0 0.0 2.0"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
                <geometry>
                    <sphere radius="0.025"/>
                </geometry>
        </collision>
    </link>
    <gazebo reference="toe_link">
        <material>Gazebo/Yellow</material>
    </gazebo>

    <gazebo>
        <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
            <robotNamespace>/</robotNamespace>
            <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
            <legacyModeNS>true</legacyModeNS>
        </plugin>
    </gazebo>

</robot>